### PR TITLE
WIP : API Call

### PR DESCRIPTION
**Pull request by benoit-cty**
This PR add the option to call the Code Carbon API:
- Add a codecarbon/core/api_client.py class to call the API
- Add codecarbon/core/schemas.py to handle object used to call API
- Many modification in codecarbon/emissions_tracker.py to handle API Call.
- The API could be call every n occurrence of _measure_power
- The cloud check is down only one time and not at every out() call.
- The CSV output could be used with the API.
- Add a delta mode to _prepare_emissions_data to send the delta to the API and not the total.
- Add a sample script examples/do_nothing.py to demo the API usage.

You can test it like this:
- Clone the repository in another folder, go in it and run:
- `docker-compose up`
- In your usual folder do:
- `git checkout api_call`
- `pip install --editable . --upgrade && python examples/do_nothing.py`
- Go to http://localhost:5080 and query the emissions table: `SELECT sum(emissions), sum(duration), sum(energy_consumed)  FROM public.emissions`

Known problems:
- If an emission API call failed, we loose the data.
- Duration is stored in second in integer, so the total do not match the real total runtime.

Works to do next :
- Handle call failure
- Unit Test

### CODE REVIEW

**Review by SaboniAmine — APPROVED** [review]

File: codecarbon/emissions_tracker.py
Comment (vict0rsch) [inline codecarbon/emissions_tracker.py:44]: I get the point but very lengthy var name... :)

File: codecarbon/emissions_tracker.py
Comment (vict0rsch) [inline codecarbon/emissions_tracker.py:227]: maybe we should overload the `-` and `+` operations so that we centralize such manipulations?

File: codecarbon/emissions_tracker.py
Comment (vict0rsch) [inline codecarbon/emissions_tracker.py:395]: currently unused, shouldn't it be passed down to the initialization of the emission trackers?

File: codecarbon/output.py
Comment (vict0rsch) [inline codecarbon/output.py:83]: I think we should keep this for people wanting to push to another system, which would mean to have another class `APIOutput` or something. Thoughts?

File: examples/do_nothing.py
Comment (vict0rsch) [inline examples/do_nothing.py:23]: That's not part of the final PR is it? It should be fixed in `logger_unique`

File: tests/test_api_call.py
Comment (benoit-cty) [reply tests/test_api_call.py:13]: Yes, I fix this.

File: codecarbon/emissions_tracker.py
Comment (vict0rsch) [reply codecarbon/emissions_tracker.py:44]: `api_call_lag` ?

Comment (vict0rsch) [reply codecarbon/emissions_tracker.py:44]: `api_call_interval`

**Review by vict0rsch — CHANGES_REQUESTED** [review]

File: codecarbon/output.py
Comment (benoit-cty) [reply codecarbon/output.py]: Will we will do it in another release. For now, the uuid is no more in the package code.

File: codecarbon/core/api_client.py
Comment (benoit-cty) [reply codecarbon/core/api_client.py:33]: It will be made in a next release.

File: codecarbon/core/schemas.py
Comment (benoit-cty) [reply codecarbon/core/schemas.py:3]: Will be done by https://github.com/mlco2/codecarbon/issues/181

File: codecarbon/emissions_tracker.py
Comment (benoit-cty) [reply codecarbon/emissions_tracker.py]: Done

File: codecarbon/emissions_tracker.py
Comment (benoit-cty) [reply codecarbon/emissions_tracker.py]: I've created a method in the class to compute the delta.

File: codecarbon/emissions_tracker.py
Comment (benoit-cty) [reply codecarbon/emissions_tracker.py]: You're right !

File: codecarbon/output.py
Comment (benoit-cty) [reply codecarbon/output.py:83]: OK, I've created a CodeCarbonAPIOutput class and leave a model for HTTPOutput.

File: examples/do_nothing.py
Comment (benoit-cty) [reply examples/do_nothing.py]: Removed !

File: codecarbon/emissions_tracker.py
Comment (benoit-cty) [inline codecarbon/emissions_tracker.py:203]: Check if we have to un comment these lines.

File: codecarbon/output.py
Comment (benoit-cty) [inline codecarbon/output.py:7]: Put it back ?

File: codecarbon/output.py
Comment (benoit-cty) [inline codecarbon/output.py:83]: Put it back in HTTPOut ?

File: codecarbon/core/api_client.py
Comment (Lothiraldan) [inline codecarbon/core/api_client.py:55]: We might want to send UTC timestamps

File: codecarbon/core/api_client.py
Comment (Lothiraldan) [inline codecarbon/core/api_client.py:93]: We might receives non-JSON response in case the URL is wrong, the server is down or a random proxy/firewall blocks the connection.

```suggestion
            f" API return http code {response.status_code} and answer : {response.text}"
```

**Review by Lothiraldan — COMMENTED** [review]
Some comments and one question, how are users supposed to use HTTPOutput after this change is merged?

**Comment by benoit-cty** [comment]
@Lothiraldan 
> Some comments and one question, how are users supposed to use HTTPOutput after this change is merged?

HTTPOutput is still there, do you now someone that use it to check it still work as expected ?

File: codecarbon/core/api_client.py
Comment (SaboniAmine) [reply codecarbon/core/api_client.py:55]: Agree, should we send local timezone identifier to let the front end app reconstruct a correct time ?
It may be stored at the run / run config level

**Comment by vict0rsch** [comment]
@benoit-cty I think BCG does ; maybe @LiamConnell can help?

**Comment by LiamConnell** [comment]
@benoit-cty @vict0rsch Still works with this 👍

File: codecarbon/core/api_client.py
Comment (benoit-cty) [reply codecarbon/core/api_client.py]: I have made a change to send: `2021-06-07T22:31:31.142202+02:00` but in the database it seems to be just in UTC : `2021-06-07T20:31:31.142202`
Is it OK for you ?

**Comment by benoit-cty** [comment]
>  Still works with this +1
@LiamConnell Well I broke HTTPOut, now it must work, can you give a try with this branch and last commit ?

**Comment by vict0rsch** [comment]
This PR LGTM just waiting for @LiamConnell's review.

@benoit-cty fo you want to update the docs in another PR?

**Comment by vict0rsch** [comment]
TODO: add timeout

File: codecarbon/core/api_client.py
Comment (Lothiraldan) [inline codecarbon/core/api_client.py:64]: Let's add a timeout here

**Review by Lothiraldan — CHANGES_REQUESTED** [review]
Let's add a

**Review by LiamConnell — APPROVED** [review]
LGTM (and all compatible with BCG's use)

File: codecarbon/core/api_client.py
Comment (benoit-cty) [reply codecarbon/core/api_client.py]: I've added a 2 seconds timeout.

**Comment by benoit-cty** [comment]
> @benoit-cty fo you want to update the docs in another PR?
@vict0rsch Doc updated.

**Comment by vict0rsch** [comment]
@benoit-cty should we also make the actual docs (rst readthedocs files) part of this PR or do them next?

**Comment by benoit-cty** [comment]
Hello, @Lothiraldan and @vict0rsch could you approve the PR ?

**Review by vict0rsch — APPROVED** [review]

**Review by Lothiraldan — APPROVED** [review]
LGTM

**Pull request merged**
